File: UnitTest/Zeze/Trans/TestNestProcedureModifyMapSet.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Zeze.Transaction;

namespace UnitTest.Zeze.Trans
{
    [TestClass]
    public class TestNestProcedureModifyMapSet
    {
        [TestInitialize]
        public void TestInit()
        {
            demo.App.Instance.Start();
        }

        [TestCleanup]
        public void TestCleanup()
        {
            demo.App.Instance.Stop();
        }

        [TestMethod]
        public void TestNestModifyMap()
        {
            demo.App.Instance.Zeze.NewProcedure(async () =>
            {
                await demo.App.Instance.demo_Module1.Table1.RemoveAsync(1);
                return 0;
            }, "ModifyMapRemove").CallAsync().Wait();

            demo.App.Instance.Zeze.NewProcedure(async () =>
            {
                var value = await demo.App.Instance.demo_Module1.Table1.GetOrAddAsync(1);
                value.Map15[1] = 1;

                await demo.App.Instance.Zeze.NewProcedure(async () =>
                {
                    var value = await demo.App.Instance.demo_Module1.Table1.GetOrAddAsync(1);
                    Assert.IsTrue(value.Map15.TryGetValue(1, out var mv1));
                    Assert.AreEqual(1, mv1);
                    value.Map15[1] = 2;
                    Assert.IsTrue(value.Map15.TryGetValue(1, out var mv2));
                    Assert.AreEqual(2, mv2);
                    return Procedure.LogicError;
                }, "ModifyMapPut2").CallAsync();

                Assert.IsTrue(value.Map15.TryGetValue(1, out var mv1));
                Assert.AreEqual(1, mv1);
                return 0;
            }, "ModifyMapPut1").CallAsync().Wait();
        }

        [TestMethod]
        public void TestNestModifySet()
        {
            demo.App.Instance.Zeze.NewProcedure(async () =>
            {
                await demo.App.Instance.demo_Module1.Table1.RemoveAsync(1);
                return 0;
            }, "ModifyMapRemove").CallAsync().Wait();

            demo.App.Instance.Zeze.NewProcedure(async () =>
            {
                var value = await demo.App.Instance.demo_Module1.Table1.GetOrAddAsync(1);
                value.Set10.Add(1);

                await demo.App.Instance.Zeze.NewProcedure(async () =>
                {
                    var value = await demo.App.Instance.demo_Module1.Table1.GetOrAddAsync(1);
                    Assert.IsTrue(value.Set10.Contains(1));
                    value.Set10.Remove(1);
                    Assert.IsFalse(value.Set10.Contains(1));
                    return Procedure.LogicError;
                }, "ModifySetRemove1").CallAsync();

                Assert.IsTrue(value.Set10.Contains(1));
                return 0;
            }, "ModifySetAdd1").CallAsync().Wait();
        }
    }
}

File: ZezeJava/ZezeJavaTest/src/UnitTest/Zeze/Trans/TestNestProcedureModifyMapSet.java
package UnitTest.Zeze.Trans;

import Zeze.Transaction.Procedure;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

public class TestNestProcedureModifyMapSet {
	@BeforeEach
	public final void testInit() throws Exception {
		demo.App.getInstance().Start();
	}

	@AfterEach
	public final void testCleanup() throws Exception {
		//demo.App.getInstance().Stop();
	}

	@Test
	public void testNestModifyMap() throws Exception {
		Assertions.assertEquals(Procedure.Success, demo.App.Instance.Zeze.newProcedure(() -> {
			demo.App.Instance.demo_Module1.getTable1().remove(1L);
			return 0L;
		}, "ModifyMapRemove").call());

		Assertions.assertEquals(Procedure.Success, demo.App.Instance.Zeze.newProcedure(() -> {
			var value = demo.App.Instance.demo_Module1.getTable1().getOrAdd(1L);
			value.getMap15().put(1L, 1L);

			Assertions.assertNotEquals(Procedure.Success, demo.App.Instance.Zeze.newProcedure(() -> {
				Assertions.assertNotNull(value.getMap15().get(1L));
				Assertions.assertEquals(1, (long)value.getMap15().get(1L));
				value.getMap15().put(1L, 2L);
				Assertions.assertEquals(2, (long)value.getMap15().get(1L));
				return Zeze.Transaction.Procedure.LogicError;
			}, "ModifyMapPut2").call());

			Assertions.assertEquals(1, (long)value.getMap15().get(1L));
			return 0L;
		}, "ModifyMapPut1").call());
	}

	@Test
	public void TestNestModifySet() throws Exception {
		Assertions.assertEquals(Procedure.Success, demo.App.Instance.Zeze.newProcedure(() -> {
			demo.App.Instance.demo_Module1.getTable1().remove(1L);
			return 0L;
		}, "ModifyMapRemove").call());

		Assertions.assertEquals(Procedure.Success, demo.App.Instance.Zeze.newProcedure(() -> {
			var value = demo.App.Instance.demo_Module1.getTable1().getOrAdd(1L);
			value.getSet10().add(1);

			Assertions.assertNotEquals(Procedure.Success, demo.App.Instance.Zeze.newProcedure(() -> {
				Assertions.assertTrue(value.getSet10().contains(1));
				value.getSet10().remove(1);
				Assertions.assertFalse(value.getSet10().contains(1));
				return Zeze.Transaction.Procedure.LogicError;
			}, "ModifySetRemove1").call());

			Assertions.assertTrue(value.getSet10().contains(1));
			return 0L;
		}, "ModifySetAdd1").call());
	}
}
